### DISCUSSION TITLE
Python: document how to manage monorepo in IDE integration

### DISCUSSION THREAD

**Discussion by helderco** [dagger/dagger#12203] [Help]
### What is the issue?

Question from user in [Discord](https://discord.com/channels/707636530424053791/1217205799684935731/1289018776863375424) about how to manage multiple virtual environments in **PyCharm**, and realized we can help with more clear instructions here:

> Try this, have PyCharm opened in your existing project, and File > Open the Dagger module’s directory. You should see a dialog with the option to **attach** to the opened project. Just select that and, assuming you already have the `.venv` inside the module directory, PyCharm should now see two interpreters.
>
> <img width="578" alt="Screenshot 2024-09-27 at 08 29 53" src="https://github.com/user-attachments/assets/4e3278bd-eab7-4aa9-86bc-1a98e2a1a6ba">
>
> If you open a file in your app, you should see the interpreter for your app’s .venv already selected. Then open a file from the module, and you should see the interpreter switch to the module’s .venv automatically. So if you jump to a `dagger` object, it should open the file in `<module source>/sdk/src/dagger/...`.

## What's next?

I don't use PyCharm or VS Code with Dagger (I use NeoVim in Warp), so I need to research (by using) the best workflow in these IDEs before making a recommendation.

## Sharing `.venv`

There's also something else I've been considering for a while: [uv workspaces](https://docs.astral.sh/uv/concepts/workspaces/). That would mean deeper integration between a Dagger module and an existing project though, sharing the same `.venv`. I don't think it's always the best option, but can be documented as a possibility too.

**Comment by jedevc** [comment]
Honestly, once you work out where in the docs this should go, and the kind of things we should cover, we should also have similar docs for go and typescript.

**Comment by ajaegle** [comment]
Attaching a project to an existing primary project might still require to set the interpreter accordingly to use the `venv` from inside the dagger directory instead of the primary project's one.

<img width="1595" alt="image" src="https://github.com/user-attachments/assets/7d65325b-aaf1-45c1-b589-d5d5a49614c5">

**Comment by ajaegle** [comment]
I did some tests with uv workspaces a week ago (outside of the dagger context) to figure out how they are supposed to be used. While dependencies are specified on the nested projects seem isolated, they are all installed in a single venv and thus must not have conflicting requirements with the other workspace packages. This could be problematic looking at what libraries are used for a minimal Dagger setup. And even if there are no conflicting requirements, I think it's still better to not pollute the outer venv with excess dependencies as this might lead to importing stuff in the main project that is only available because it's a dependency of Dagger itself.

Also, even though uv is gaining a lot of traction, not everyone adopted it (yet). This would require different setups for uv and non-uv root projects, right?

The UV docs also list some points on why workspaces might not be a valid choice: https://docs.astral.sh/uv/concepts/workspaces/#when-not-to-use-workspaces

You can find my (minimal) experiment here: https://github.com/ajaegle/python-uv-workspaces-eval

**Comment by helderco** [comment]
I know how workspaces work, the Python SDK uses it: 🙂

https://github.com/dagger/dagger/blob/d12a6f38adc24049287d50bf5d40ee85fd2d9666/sdk/python/pyproject.toml#L69-L70

Setting up the workspace won't be something that Dagger will manage for you. I'm just talking about documenting how this can be done for those interested in this solution, but it's up to the user to set it up.

**Comment by Siafu** [comment]
I have a decent environment setup using vscode. If anyone is interested, I can demo. It's set up using nx monorepo and a view vscode plugins. I could do a video on this if anyone is interested.

**Comment by TomChv** [comment]
> I have a decent environment setup using vscode. If anyone is interested, I can demo. It's set up using nx monorepo and a view vscode plugins. I could do a video on this if anyone is interested.

This is definitely interesting, is there something different in your setup than what we explain there: https://docs.dagger.io/manuals/developer/ide-integration ?

**Comment by danielgafni** [comment]
I'm new to `dagger` but ain't new to uv workspaces. 

I currently have a monorepo with uv workspaces set up. 

I'm kinda struggling to create `dagger` projects in this setup:
- I'd like to have these projects as workspaces members. This means they won't get their own `uv.lock` files (all members share a single `.venv` a `uv.lock` file in the repo root). it's unclear how to override the default containerized `dagger` env build by using the repo root `uv.lock` file.
- It's unclear **why** is `dagger` even vendoring itself via the `sdk` directory. Why can't I just include `dagger-io` as a normal dependency installed from PyPI? In fact, I tried doing that and it seems to be working... For now. 

I decided not to include my Dagger projects as a workspace member yet (otherwise `dagger` fails to build it without the expected `uv.lock` file). I can install it as a normal `path` dependency to but that's not ideal. 

Are there any better options available?

**Comment by helderco** [comment]
Hi, @danielgafni! 

> * It's unclear **why** is `dagger` even vendoring itself via the `sdk` directory. Why can't I just include `dagger-io` as a normal dependency installed from PyPI? In fact, I tried doing that and it seems to be working... For now.

Vendoring was a quick solution when we were working on releasing Dagger modules. We're [working towards doing what you described](https://github.com/dagger/dagger/pull/9538), but the reason for vendoring is because of the custom client bindings.

If you don’t have [module dependencies](https://docs.dagger.io/api/module-dependencies) then it’ll be the same as the published `dagger-io` package in PyPI (i.e., only client bindings for the core API), so that’s why it’s working for you. But if you add a dependency (`dagger install`), then you won’t see it in the `dag` object.

> I decided not to include my Dagger projects as a workspace member yet (otherwise `dagger` fails to build it without the expected `uv.lock` file). I can install it as a normal `path` dependency to but that's not ideal.
> 
> Are there any better options available?

This is something the Python SDK needs to support as an advanced use case, but it needs a few fixes because the SDK makes a few assumptions atm, particularly when installing dependencies, because it checks for the existence of a `uv.lock` file vs a `requirements.lock` file to use trigger uv [project management](https://docs.astral.sh/uv/guides/projects/) behavior or [pip interface](https://docs.astral.sh/uv/pip/) (legacy).

You can help figure out how best to do this and I’ll make sure to add support for it.

There are a few things that can help to understand first:
- Your module runs in it's own [ephemeral container](https://docs.dagger.io/api/module-structure#runtime-container)
- Your module’s files are mounted in that container, starting from your repo root, but doesn’t include any files outside of it by default.
- When there’s no `uv.lock` file in the module, the SDK creates a new one with [`uv lock`](https://github.com/dagger/dagger/blob/717788d820fed40deeb8f8476e21627352926bfc/sdk/python/runtime/main.go#L376) so we need to change this somehow.
- When installing dependencies, if there’s a `uv.lock` file it runs [`uv sync --no-dev`](https://github.com/dagger/dagger/blob/717788d820fed40deeb8f8476e21627352926bfc/sdk/python/runtime/main.go#L407). Does that command work for you (assuming shared workspace) from the module’s directory?
- The SDK manages a virtual env in `/opt/venv` when using `uv sync`, because it was easier than running the entrypoint with `uv run`. Not sure if it needs changing.

### 1. Include needed files

So, the first thing to do is edit `dagger.json` and [explicitly include](https://docs.dagger.io/configuration/modules#file-and-directory-filters) any files that uv needs for picking up the parent project and build successfully. For example:
```json
  "include": [
    "../pyproject.toml",
    "../uv.lock"
  ]
```

Don’t add `.venv`. It won’t be good for caching, but also won’t work because of the global `/opt/venv`. I don’t think it matters though, and uv’s cache should be persisted between runs (best effort). 

I think you’ll pull in all your root project’s dependencies

### 2. Check filters

 You can check if your filters are working as expected by inspecting the entire contents of what’s mounted into the container with:
 ```console
 dagger core module-source --ref-string . resolve-from-caller context-directory glob --pattern '**'
 ```
 Or open a shell on it to navigate and inspect:
 ```console
 dagger core module-source --ref-string . resolve-from-caller context-directory terminal
 ```
 That’s assuming you’re in the same dir as your module’s `dagger.json`, otherwise adjust the argument to `--ref-string`.
 
### 3. Let’s figure it out

What I need to do next is to figure out the right commands for installing dependencies and running the entrypoint. Not sure if those need changing.

With `uv.lock`, here’s what I’m thinking:
- There’s a `uv.lock`? → Run `uv lock`
- No `uv.lock` and no `requirements.lock`? → ~~Run `uv lock`~~ **do nothing**
- No `uv.lock` but there’s a `requirements.lock`? → Use *pip interface*

Then I need to run `uv sync` when there's no `uv.lock` and no `requirements.lock` too, not just when there's a `uv.lock`. I suppose that `uv sync` will pickup the parent’s `pyproject.toml` and `uv.lock` then. Need to try it.

As a workaround in the meantime, if you can't wait, you can [extend the SDK](https://github.com/dagger/dagger/tree/main/core/integration/testdata/modules/python/extended) or fork it.

**Comment by danielgafni** [comment]
Hey @helderco, big thanks for the detailed reply, I appreciate it! 

> There are a few things that can help to understand first:
> ...

~~Thanks, this matches with my assumptions. Is there a way to modify/override these steps? Can I roll out my own build for the module?~~ Your (3) point addresses this 

> So, the first thing to do is edit dagger.json and [explicitly include](https://docs.dagger.io/configuration/modules#file-and-directory-filters) 

Great hint. I might make use of this option. 

---

Now, re: Dagger improvements. 

> No uv.lock and no requirements.lock? → Run uv lock do nothing

I think this shouldn't really happen. It's not like `uv.lock` is actually missing in my project. It's just in a different location. 

I should be able to specify the correct context for running the module build (e.g. repo root) and the path to the module source code. The module build process should be: 
- copy `uv.lock` and `pyproject.toml` (I think it's required for workspaces to work) from the repo root
- run `uv sync --no-install-workspace --no-dev --package <sdk-package-name>` (this way only third party deps are installed and cached in this step)
- copy module source code from the specified location **to the same relative location** inside the build container - this way it will be recognized by `uv`
- run `uv sync --no-dev --package <sdk-package-name>`


For reference, here is a piece from my typical `Dockerfile` used to build `uv` workspace members:

```Dockerfile

WORKDIR /src

COPY pyproject.toml uv.lock  ./

ARG PACKAGE
ARG PACKAGE_DIR

# I don't remember if this is strictly required 
COPY $PACKAGE_DIR/pyproject.toml ./$PACKAGE_DIR/

RUN --mount=type=cache,target=/root/.cache/uv \
    uv sync --all-extras --no-dev --no-install-workspace --package $PACKAGE

COPY $PACKAGE_DIR ./$PACKAGE_DIR

ARG PACKAGE

# finally install all our code
RUN uv sync --all-extras --no-dev --package $PACKAGE
```

**Comment by helderco** [comment]
Thanks for the Dockerfile!


> > No uv.lock and no requirements.lock? → Run uv lock do nothing
> 
> I think this shouldn't really happen. It's not like `uv.lock` is actually missing in my project. It's just in a different location.
> 
> I should be able to specify the correct context for running the module build (e.g. repo root) and the path to the module source code. The module build process should be:

What I meant is that you can `uv sync` a package in it's own subdir, and uv will detect that it's part of a workspace, and use the root's .venv, etc...

For example, the Python SDK has a [`codegen` subpackage](https://github.com/dagger/dagger/tree/f2fcc7ffc8394ef4a13c4aebb66ca45f3bea7491/sdk/python/codegen):

https://github.com/dagger/dagger/blob/f2fcc7ffc8394ef4a13c4aebb66ca45f3bea7491/sdk/python/pyproject.toml#L66-L70

You can see that package doesn't have a `uv.lock`, but I can run `uv sync` from that directory. Notice the paths:

<pre>
sdk/python/codegen ❯ uv sync --verbose
DEBUG uv 0.5.30 (ddbc6e315 2025-02-10)
DEBUG <strong>Found project root: `/Users/helder/Projects/dagger/sdk/python/codegen`</strong>
DEBUG <strong>Found workspace root: `/Users/helder/Projects/dagger/sdk/python`</strong>
DEBUG Adding root workspace member: `/Users/helder/Projects/dagger/sdk/python`
DEBUG Adding current workspace member: `/Users/helder/Projects/dagger/sdk/python/codegen`
DEBUG Acquired lock for `/Users/helder/Projects/dagger/sdk/python`
DEBUG Reading Python requests from version file at `/Users/helder/Projects/dagger/sdk/python/.python-version`
DEBUG Using Python request `3.12` from version file at `/Users/helder/Projects/dagger/sdk/python/.python-version`
DEBUG <strong>Checking for Python environment at `/Users/helder/Projects/dagger/sdk/python/.venv`</strong>
DEBUG The virtual environment's Python version satisfies `3.12`
DEBUG Released lock at `/var/folders/qh/fcm1jq4j3p9d2js8_k_cbsyh0000gn/T/uv-c6731e89678276d8.lock`
...
DEBUG <strong>Existing `uv.lock` satisfies workspace requirements</strong>
</pre>


> * copy `uv.lock` and `pyproject.toml` (I think it's required for workspaces to work) from the repo root

This is done by the added inclusions in `dagger.json`.

> * run `uv sync --no-install-workspace --no-dev --package <sdk-package-name>` (this way only third party deps are installed and cached in this step)

I think `uv sync --no-install-workspace --no-dev` is harmless. The `--no-install-workspace` should be ignored when not in a workspace:

```
❯ uv sync --no-dev --no-install-workspace --verbose
DEBUG uv 0.5.30 (ddbc6e315 2025-02-10)
DEBUG Found project root: `/Users/helder/Projects/dagger/sdk/python/dev`
DEBUG Found workspace root `/Users/helder/Projects/dagger/sdk/python`, but project is not included
DEBUG No workspace root found, using project root
```

And the `--package <sdk-package-name>` isn't necessary if you're inside that package's directory.

> * copy module source code from the specified location **to the same relative location** inside the build container - this way it will be recognized by `uv`

This is how mounting sources work for all SDKs already.

> # finally install all our code
> RUN uv sync --all-extras --no-dev --package $PACKAGE

Do you need the `--all-extras`?

---

I really want to avoid having to customize the `uv sync` command as that would open it too much and hard to support. However, we're planning on allowing users to set env vars in `dagger.json`, and get them set in the module's container. This will allow you to configure a lot from uv.

That's why I think it's best to keep the uv commands as minimal as possible so you can configure with env vars, or some settings through `pyproject.toml` or `uv.toml`.

**Comment by danielgafni** [comment]
> You can see that package doesn't have a uv.lock, but I can run uv sync from that directory. Notice the paths:

Oh ok, amazing. So we just need to copy `uv.lock` to the container correctly. 

> `--all-extras` honestly, not sure about it. It's kinda hard to imagine extras for a dagger module since there is typically just variant of the module packaged anyway. 

I suggested additional steps for better caching because I suspected something was wrong about it: it sometimes takes up to 15 seconds to initialize the module after source code changes. 

```
✔ connect 0.2s
✔ load module 14.5s
✔ parsing command line arguments 0.0s
```

**Comment by slamer59** [comment]
Hello, 
thanks for adressing this. It is definitely a pretty dense subject ! 

You seems exprets to uv monorepo. Can we share good documentation on this : challenges, etc ? 

To put in my use case : 
- I am using uv as package manager (so as dagger internally) 
- I am using pydantic settings (but it can be anything else) for my application
- I want to use this class with dagger. 

Here I find a simple exemple
- https://github.com/JasperHG90/uv-monorepo : with uv
- https://github.com/tweag/python-monorepo-example : with uv and reqs

**Comment by danielgafni** [comment]
Hi, I just made a blog post about using `uv` with `Dagger`: https://gafni.dev/blog/cracking-the-python-monorepo/

It should give you a reference setup for the potential Dagger pipelines with `uv`.
